modules/shared/src/main/scala/org/tessellation/currency/dataApplication/package.scala

```diff
@@ -87,6 +87,8 @@ trait BaseDataApplicationL0Service[F[_]] extends BaseDataApplicationService[F] w
   def genesis: DataState.Base
 
   final def serializedOnChainGenesis: F[Array[Byte]] = serializeState(genesis.onChain)
+
+  def onSnapshotAcceptedLive(snapshot: Signed[CurrencyIncrementalSnapshot]): F[Unit]
```

Comment:
IMO this function should be optional to be implemented, with fallback to `noop` operation

Reply:
@marcinwadon Where do you suggest the default implementation go?

Reply:
what do you mean @sdavidp ?

Reply:
you mean this?

```scala
def onSnapshotAcceptedLive(snapshot: Signed[CurrencyIncrementalSnapshot]): F[Unit] = Applicative[F].unit
```

Reply:
@marcinwadon Yes but I can't do that because `Applicative` is not available to me on the trait.

Reply:
Not sure if I understand correctly but for me we can require the user to supply this implementation and it can be a simple unit like in the example above if nothing needs to be done which is an equivalent of noop.

Reply:
@TheMMaciek That was my expectation as well so I don't understand the original comment. @marcinwadon Can you please be more specific?

Reply:
Yep. We may specific more "handlers", like:
- onSnapshotAcceptance,
- onConsumeSnapshot,
- onArtifactCreated,
and so on.

They all should be optional to be implemented by the end user. One option would be to specify them as Option but it does not make sense for me. It's better to deliver default implementation "Noop", and if anyone wants to bind to the handler, one can provide own implementation. It does make not sense to require end user to provide all implementations for all the handlers. The easiest way would be to require the Applicative (can be directly for that function) which specifies "pure" method. For someone using IO it would be IO[Unit], for someone else it may be Id so just ().

Reply:
@marcinwadon Yes I understand all that. My question is where do you want the default _noop_ implementation to live? Context bounds are not permitted in a `trait` so it can't be there. The only options I can think of is to turn one of the traits to an abstract class or to provide the context bounds on an _apply_.

Reply:
Not apply but Applicative. You can add it directly to the function that requires it, so the one you've just added